Block volumes Support: FC plugin update

mtanino opened the pull request:
**What this PR does / why we need it**:

Add interface changes to FC volume plugin to enable block volumes support feature.

**Which issue this PR fixes**: 
Based on this proposal (kubernetes/community#805 & kubernetes/community#1265) and this feature issue: kubernetes/features#351

**Special notes for your reviewer**:

This PR temporarily includes following changes except FC plugin change for reviewing purpose.
These changes will be removed from the PR once they are merged.
- (#50457) API Change 
- (#53385) VolumeMode PV-PVC Binding change
- (#51494) Container runtime interface change, volumemanager changes, operationexecutor changes

There are another PRs related to this functionality.
(#50457) API Change
(#53385) VolumeMode PV-PVC Binding change
(#51494) Container runtime interface change, volumemanager changes, operationexecutor changes
(#55112) Block volume: Command line printer update
Plugins
(#51493) Block volumes Support: FC plugin update
(#54752) Block volumes Support: iSCSI plugin update

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```
FC plugin: Support for block volume 
- This enables uses to allow attaching raw block volume to their pod without filesystem through FC plugin.
```

mtanino commented:
/cc @msau42 @liggitt @jsafrane @saad-ali @erinboyd @screeley44

jsafrane commented on pkg/volume/fc/disk_manager.go:
```diff
@@ -117,5 +119,48 @@ func diskTearDown(manager diskManager, c fcDiskUnmounter, volPath string, mounte
+	if !filepath.IsAbs(globalPDPath) || !filepath.IsAbs(volPath) {
+		return fmt.Errorf("These paths should be absolute: globalPDPath: %s, volPath: %s", globalPDPath, volPath)
+	}
+	devicePath, err := os.Readlink(globalPDPath + "/" + "symlink")
```
Who creates this symlink and when? Who deletes it?

mtanino replied:
@jsafrane Under discussion on the spec.

jsafrane commented on pkg/volume/fc/disk_manager.go:
```diff
@@ -117,5 +119,48 @@ func diskTearDown(manager diskManager, c fcDiskUnmounter, volPath string, mounte
+		glog.Errorf("failed to readlink: %s", globalPDPath+"/"+"symlink")
+		return err
+	}
+	noMnt, err := b.mounter.IsLikelyNotMountPoint(volPath)
```
Why this check? IMO it can't be ever a mount point

mtanino replied:
@jsafrane I supposed there is a case that the volume is previously used as filesystem and remained as mounted based on unmount failure. Doesn't this need to consider?

msau42 replied:
If it was mounted as a filesystem, the pod mount would be under the volumes directory and not the devices directory.

mtanino replied:
That's right! will remove.

jsafrane commented on pkg/volume/fc/disk_manager.go:
```diff
@@ -117,5 +119,48 @@ func diskTearDown(manager diskManager, c fcDiskUnmounter, volPath string, mounte
+		return err
+	}
+	if fi.Mode()&os.ModeSymlink == os.ModeSymlink {
+		err := os.Remove(volPath + "/" + "symlink")
```
IMO, it should remove also the directory `volPath`. Or who deletes it and when?

mtanino replied:
ack.

jsafrane commented on pkg/volume/fc/fc.go:
```diff
@@ -249,6 +320,77 @@ func (c *fcDiskUnmounter) TearDownAt(dir string) error {
+	return devicePath
+}
+
+func (b *fcDiskMapper) GetVolumeMode() v1.PersistentVolumeMode {
```
This method does not seem to be called.

mtanino replied:
@jsafrane This method is currently called from makeBlockVolumes() in kubelet_pod.go but is under discussion on the other thread. May be we can remove.

https://github.com/kubernetes/kubernetes/pull/51494

jsafrane replied:
Do you have a branch or complete PR with all the code that would compile and run? Cross-reviewing several partial PRs is hard.

mtanino replied:
@jsafrane 
We are developing this feature in parallel, therefore we don't have complete PR.

But let me try something to review easily.
I understand these separated PRs are too hard to review without compile and run them.

msau42 replied:
You can cherry-pick the other PRs into your branch temporarily, and then remove them when they get merged.

mtanino replied:
@msau42 @jsafrane 
yeah, good idea. will do.

jsafrane commented on pkg/volume/fc/fc.go:
```diff
@@ -185,15 +249,22 @@ type fcDisk struct {
 func (fc *fcDisk) GetPath() string {
 	name := fcPluginName
-	// safe to use PodVolumeDir now: volume teardown occurs before pod is cleaned up
+	// safe to use PodppingVolumeDir now: volume teardown occurs before pod is cleaned up
```
typo?

mtanino replied:
ack.

jsafrane reviewed (changes requested):
IMO the whole device cleanup/detach needs more thoughts. I commented on the proposal

mtanino commented on pkg/volume/fc/attacher.go:
```diff
@@ -66,8 +66,8 @@ func (attacher *fcAttacher) VolumesAreAttached(specs []*volume.Spec, nodeName ty
 
-func (attacher *fcAttacher) WaitForAttach(spec *volume.Spec, devicePath string, _ *v1.Pod, timeout time.Duration) (string, error) {
-	mounter, err := volumeSpecToMounter(spec, attacher.host)
+func (attacher *fcAttacher) WaitForAttach(spec *volume.Spec, devicePath string, pod *v1.Pod, timeout time.Duration) (string, error) {
```
@jsafrane I understand. let me investigate another way.

Commit:
Block volumes Support: FC plugin update

This patch adds block volume support to FC volume plugin.

jsafrane commented on pkg/volume/fc/attacher.go:
```diff
@@ -160,7 +161,7 @@ func (detacher *fcDetacher) UnmountDevice(deviceMountPath string) error {
 }
 
-func volumeSpecToMounter(spec *volume.Spec, host volume.VolumeHost) (*fcDiskMounter, error) {
+func volumeSpecToMounter(spec *volume.Spec, host volume.VolumeHost, pod *v1.Pod) (*fcDiskMounter, error) {
```
`pod` argument is not used. Can it be removed?

mtanino replied:
Done. pod is removed from argument.

jsafrane commented on pkg/volume/fc/disk_manager.go:
```diff
@@ -117,5 +119,44 @@ func diskTearDown(manager diskManager, c fcDiskUnmounter, volPath string, mounte
+		glog.Errorf("failed to readlink: %s", globalPDPath+"/"+"symlink")
+		return err
+	}
+	if err != nil && !os.IsNotExist(err) {
```
this check is always false, there is `if err != nil ` three lines above.

mtanino replied:
This logic was removed

jsafrane commented on pkg/volume/fc/fc.go:
```diff
@@ -126,17 +126,60 @@ func (plugin *fcPlugin) newMounterInternal(spec *volume.Spec, podUID types.UID,
+			io:      &osIOHandler{},
+			plugin:  plugin},
+		fsType:     fc.FSType,
+		volumeMode: "",
```
Why is volumeMode empty?

mtanino replied:
Set v1.PersistentVolumeFilesystem to volumeMode .

jsafrane commented on pkg/volume/fc/fc.go:
```diff
@@ -126,17 +126,60 @@ func (plugin *fcPlugin) newMounterInternal(spec *volume.Spec, podUID types.UID,
+
+	var lun string
+	var wwids []string
+	if fc.Lun != nil && len(fc.TargetWWNs) != 0 {
```
please extract a function that computes wwids, I see that NewMounter has the same code.

mtanino replied:
Done.

jsafrane commented on pkg/volume/volume.go:
```diff
@@ -132,6 +132,28 @@ type Unmounter interface {
+	//
+	SetUpDeviceAt(dir string) error
+	//
+	GetVolumePath() string
```
Should it be `GetDevicePath`?

mtanino replied:
Agreed.

mtanino commented:
Updated to align latest interface update. Also comments from @jsafrane were addressed.

TODO:
- Add comments and remove debug code
- Add more error handling code
- Add test code, test changes

msau42 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -627,6 +627,262 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+		}
+
+		// Map device to volume map path under the given pod directory using symbolic link
+		// A plugin doesn't have attacher also needs to map device to global map path with SetUpDevice()
```
It's confusing that we have different requirements for SetUp depending on the plugin.  We should have consistent requirements for implementing the interface.

mtanino replied:
OK.
If we should have consistent requirement to SetUpDevice(), the method should create symbolic link on both global map path and volume map path regardless of whether the plugin have attacher or not.
Is this make sense? 
Then we will remove MapDevice() and UnmapDevice() from operation_executor.

msau42 replied:
Would it work if we made a symlink from pod's volume directory to the global map, after calling SetUp?  Then all plugins only have to take care of setting up the volume at the pod's path.

mtanino replied:
If the SetUpDevice only take care of volume setting up, I suppose GenerateMapVolumeFunc() needs following steps,

1. SetUpDevice() -> this should return block device path
2. Create global map path and symlink to block device.
3. Create pod's volume directory and symlink from pod's volume directory to block device(or global map?)

mtanino replied:
or may be this also work.

1. SetUpDevice() -> only take care of volume setup, return error status.
2. Get block device path from mapper.GetDevicePath() 
3. Create global map path directory and symlink to block device.
4. Create pod's volume directory and symlink from pod's volume directory to block device(or global map?)

msau42 replied:
I think that can work.  It seems like the SetUp is becoming more of a "local attach" method.  For attachable plugins like gce pd, I think SetUp will be a noop since it is already attached at a global path.

mtanino replied:
>>  I think SetUp will be a noop since it is already attached at a global path.
yes, I haven't pushed my latest code, but setUp and tearDown now only do returning nil if plugin is attachable plugin.

msau42 commented on pkg/volume/util/operationexecutor/operation_executor.go:
```diff
@@ -105,6 +105,26 @@ type OperationExecutor interface {
+
+	// UnmapVolume unmaps symbolic link to the volume from both the pod specified in
+	// volumeToUnmount and global map path. And then, updates the actual state of the world to reflect that.
+	UnmapVolume(volumeToUnmount MountedVolume, actualStateOfWorld ActualStateOfWorldMounterUpdater) error
```
Can we get Map and Unmap methods to be symmetrical?

mtanino replied:
Which part do you intend to make symmetrical? Operations in those method?

msau42 replied:
I mean there's one Map function, and two Unmap functions.

mtanino replied:
UnmapVolume and UnmapDevice has different purpose and also timing which is called from reconciler is different.

- UnmapVolume is called after a volume is unmounted status
- UnmapDevice is called after a volume is detached/unmounted status

Can we combine them?

For the filesystem case, we have
- MountVolume   -> call SetUp(w/o attacher), call attacher.MountDevice and SetUp(w/ attacher)
- UnmountVolume -> call TearDown
- UnmountDevice -> ref count of mount point, call attacher.UnmountDevice and check device status if it is opened

And for the block volume case, we will have;
- MapVolume   -> call SetUpDevice(w/o attacher), call MapDevice and SetUpDevice(w/ attacher)
- UnmapVolume -> call TearDownDevice(remove symlinkc from volume map path) and
                            call UnmapDevice(remove symlinkc from global map path)
- UnmapDevice -> ref count of mount point and check device status if it is opened

msau42 replied:
Ah ok, that makes more sense then.

mtanino commented on pkg/controller/volume/attachdetach/cache/actual_state_of_world.go:
```diff
@@ -298,6 +301,7 @@ func (asw *actualStateOfWorld) AddVolumeNode(
 			spec:            volumeSpec,
 			nodesAttachedTo: make(map[types.NodeName]nodeAttachedTo),
 			devicePath:      devicePath,
+			pod:             pod,
```
@msau42 
I understand. I'll remove and find alternative way.

mtanino replied:
done.

mtanino commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -171,10 +172,22 @@ func (rc *reconciler) reconcile() {
+					mountedVolume.MountedVolume, rc.actualStateOfWorld)
+				glog.Infof("#### DEBUG LOG ####: reconcile mountedVolume.BlockVolumeMapper case")
+				//glog.Infof("#### DEBUG LOG ####: reconcile mountedVolume.BlockVolumeMapper: %s", mountedVolume.BlockVolumeMapper)
+			}
```
I'll add error case.

mtanino commented:
@jsafrane @msau42 @erinboyd @screeley44 

I've done following changes.
I believe most of implementations are done except "volumeDevices{} and PVC combination check"

Please take a look and give feedbacks.
Also this PR contains a little old api change code, because latest one can't be compiled.

`Done:`

- Add symbolic link refcount logic, Move out symbolic link creation and deletion logic from plugin
- FC plugin update based on new symbolic link logic
- Update syncStates() and reconstruct() logic for block volume
- Remove pod parameter from attachedVolume

`TODO:`
- volumeDevices{} and PVC combination check(still under consideration the place to implement it)
- Add comments and remove debug code
- Add more error handling code
- Add test code, test changes

`Other PRs:`
- API change: https://github.com/kubernetes/kubernetes/pull/50457
- Block volumes Support: CRI, volumemanager and operationexecutor changes
  Currently these changes are included on this PR.
  (https://github.com/kubernetes/kubernetes/pull/51494)
- pv-controller changes
  PR hasn't opened yet. @screeley44 will open it.
- Command Line changes
  PR hasn't opened yet. @erinboyd will open it.

mtanino commented:
@msau42 @jsafrane @erinboyd @screeley44
All comments from @msau42 are addressed. PTAL.

TODO:

- Remove debug code
- Add test code, test changes

mtanino commented:
/cc @rootfs 
PTAL.

rootfs commented on pkg/volume/fc/fc.go:
```diff
@@ -256,3 +381,19 @@ func getVolumeSource(spec *volume.Spec) (*v1.FCVolumeSource, bool, error) {
 	return nil, false, fmt.Errorf("Spec does not reference a FibreChannel volume type")
 }
+
+func getWwnsLunWwids(fc *v1.FCVolumeSource) ([]string, string, []string, error) {
```
unit test this

rootfs commented on pkg/volume/fc/fc.go:
```diff
@@ -246,6 +310,67 @@ func (c *fcDiskUnmounter) TearDownAt(dir string) error {
+	return nil
+}
+
+// Return:
```
format as godoc (and others too)

rootfs commented on pkg/volume/fc/fc.go:
```diff
@@ -104,40 +104,71 @@ func (plugin *fcPlugin) GetAccessModes() []v1.PersistentVolumeAccessMode {
+			io:      &osIOHandler{},
+			plugin:  plugin},
+		fsType:     fc.FSType,
+		volumeMode: v1.PersistentVolumeFilesystem,
```
why set to Filesystem?

mtanino replied:
I think we should get correct volumeMode from Spec.

rootfs commented on pkg/volume/fc/fc.go:
```diff
@@ -104,40 +104,71 @@ func (plugin *fcPlugin) GetAccessModes() []v1.PersistentVolumeAccessMode {
 }
 
-func (plugin *fcPlugin) newMounterInternal(spec *volume.Spec, podUID types.UID, manager diskManager, mounter mount.Interface, exec mount.Exec) (volume.Mounter, error) {
+func (plugin *fcPlugin) newMounterInternal(spec *volume.Spec, pod *v1.Pod, manager diskManager, mounter mount.Interface, exec mount.Exec) (volume.Mounter, error) {
```
is there any other use for `pod` other than getting the `UID`?

rootfs commented on pkg/volume/fc/fc.go:
```diff
@@ -104,40 +104,71 @@ func (plugin *fcPlugin) GetAccessModes() []v1.PersistentVolumeAccessMode {
+	}
+
+	var uid types.UID
+	if pod != nil {
```
when will pod be nil?

rootfs commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -91,9 +92,9 @@ func (kl *Kubelet) GetActivePods() []*v1.Pod {
+// makeGPUDevices determines the devices for the given container.
 // Experimental.
-func (kl *Kubelet) makeDevices(pod *v1.Pod, container *v1.Container) ([]kubecontainer.DeviceInfo, error) {
+func (kl *Kubelet) makeGPUDevices(pod *v1.Pod, container *v1.Container) ([]kubecontainer.DeviceInfo, error) {
```
is this change necessary?

rootfs commented on pkg/volume/util/util.go:
```diff
@@ -270,3 +270,119 @@ func stringToSet(str, delimiter string) (sets.String, error) {
+			return err
+		}
+	}
+	if err := os.Symlink(devicePath, linkPath); err != nil {
```
what about the mode of the symlink?

mtanino replied:
We talked about this on the slack, and we confirmed we don't need to care mode of symlink here.

rootfs commented:
change DEBUG LOG to glog.V(5 or higher).Infof

mtanino commented:
@jsafrane @rootfs @msau42 
Addressed comments from @rootfs.

mtanino commented:
Review of volumemanager is on going at #51494

mtanino commented:
Done:
- Add constructBlockVolume() for block specific volume reconstruction

mtanino commented:
Done:

- TearDownDevice() is moved from BlockVolumeUnmapper to BlockVolumeMapper

mtanino commented:
@msau42 @jsafrane @rootfs @erinboyd @screeley44

I think I have addressed all feedback. PTAL.
Thanks.

mtanino commented:
Done
- Update GetGlobalMapPath and GetPodDeviceMapPath
- Remove unused code

mtanino commented:
/retest

jsafrane commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -627,6 +628,316 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+
+		// The block volume is not referenced from Pods. Release file descriptor lock.
+		glog.V(5).Infof("UnmapDevice: deviceToDetach.DevicePath: %v", deviceToDetach.DevicePath)
+		exec := og.volumePluginMgr.Host.GetExec(blockVolumePlugin.GetPluginName())
```
This requires that losetup is in all containers with mount utilities. You probably want to execute the command on the host (or in the same container where is kubelet), hece use directly os/exec. And you probably don't need to pass the exec interface around, GetLoopDevice/RemoveLoopDevice et al can call exec directly.

This applies also to AttachFileDevice above.

jsafrane commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -627,6 +628,316 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+			return deviceToDetach.GenerateErrorDetailed("UnmapDevice failed", err)
+		}
+		// The globalMapPath directory is empty. Remove the directory
+		removeMapPathErr := util.RemoveMapPath(globalMapPath)
```
I am not sure about the order here. If kubelet crashes after removing the directory and before detaching the volume you lost the evidence that the volume is attached and new kubelet won't detach it. I think that the directory should be removed after detaching.

mtanino commented on pkg/volume/fc/fc.go:
```diff
@@ -171,6 +225,63 @@ func (plugin *fcPlugin) ConstructVolumeSpec(volumeName, mountPath string) (*volu
+
+	// Create volume from wwn+lun or wwid
+	var fcVolume *v1.Volume
+	if strings.Contains(volumeInfo, "-lun-") {
```
If vendor implement WWID with following NAA identifier, it's 64 or 128 bit hexadecimal value. Therefore it doesn't contain dash or lun. (https://tools.ietf.org/html/rfc3980#ref-FC-FS)

mtanino commented on pkg/volume/util/util.go:
```diff
@@ -270,3 +278,238 @@ func stringToSet(str, delimiter string) (sets.String, error) {
+	return nil
+}
+
+// IsSymlinkExist retruns true if specified symbolic link exists
```
ack.

mtanino commented:
/retest

mtanino commented:
Updates based on @jsafrane 's review.

- Change first argument of ConstructBlockVolumeSpec from podName to podUID
- Add arguments mapPath and devicePath to TearDownDevice()
- Remove to pass exec from loopdevice operation methods
- Update IsSymlinkExist()
- Remove unnecessary symlink check before removing file
- Test: add volumetesting.VerifyGetPodDeviceMapPathCallCount check
- Fix typos

mtanino commented:
/retest

mtanino commented:
/retest

mtanino commented:
@msau42 @jsafrane @rootfs 
This PR is ready to review. PTAL

krzyzacy commented:
/retest

mtanino commented:
@msau42 @jsafrane
Updated volumemanager's code/test cases based on @msau42's review. PTAL

mtanino commented:
/retest

mtanino commented:
/retest

mtanino commented:
/retest

mtanino commented:
/retest

mtanino commented:
@rootfs 
PTAL FC Plugin change.

rootfs commented on pkg/volume/fc/fc.go:
```diff
@@ -115,29 +117,64 @@ func (plugin *fcPlugin) newMounterInternal(spec *volume.Spec, podUID types.UID,
+}
+
+func (plugin *fcPlugin) NewBlockVolumeMapper(spec *volume.Spec, pod *v1.Pod, _ volume.VolumeOptions) (volume.BlockVolumeMapper, error) {
+	// Inject real implementations here, test through the internal function.
```
this comment is not relevant any more

mtanino replied:
ack. removed.

rootfs commented on pkg/volume/fc/fc.go:
```diff
@@ -115,29 +117,64 @@ func (plugin *fcPlugin) newMounterInternal(spec *volume.Spec, podUID types.UID,
+}
+
+func (plugin *fcPlugin) newBlockVolumeMapperInternal(spec *volume.Spec, podUID types.UID, manager diskManager, mounter mount.Interface, exec mount.Exec) (volume.BlockVolumeMapper, error) {
+	// fc volumes used directly in a pod have a ReadOnly flag set by the pod author.
```
these comments can go into `getVolumeSource`

mtanino replied:
done.

rootfs commented on pkg/volume/fc/fc.go:
```diff
@@ -115,29 +117,64 @@ func (plugin *fcPlugin) newMounterInternal(spec *volume.Spec, podUID types.UID,
+
+	wwns, lun, wwids, err := getWwnsLunWwids(fc)
+	if err != nil {
+		return nil, fmt.Errorf("fc: no fc disk information found. failed to make a new mounter")
```
-> new mapper

mtanino replied:
done.

rootfs commented on pkg/volume/fc/fc.go:
```diff
@@ -161,6 +198,23 @@ func (plugin *fcPlugin) newUnmounterInternal(volName string, podUID types.UID, m
 }
 
+func (plugin *fcPlugin) NewBlockVolumeUnmapper(volName string, podUID types.UID) (volume.BlockVolumeUnmapper, error) {
+	// Inject real implementations here, test through the internal function.
```
fix the comment too

mtanino replied:
done.

rootfs commented on pkg/volume/fc/fc.go:
```diff
@@ -171,6 +225,57 @@ func (plugin *fcPlugin) ConstructVolumeSpec(volumeName, mountPath string) (*volu
+	if err != nil {
+		// return dummy fc volumesource
+		fcVolume := createVolumeFromFCVolumeSource(volumeName, v1.FCVolumeSource{})
+		return volume.NewSpecFromVolume(fcVolume), err
```
why still returning a spec? can just return nil?

mtanino replied:
my bad. this should return nil, err.

rootfs commented on pkg/volume/fc/fc_util.go:
```diff
@@ -198,25 +213,31 @@ func searchDisk(b fcDiskMounter) (string, string) {
 	}
+	// If the volumeMode is 'Block', plugin don't have to format the volume.
+	// The globalPDPath will be created by operationexecutor. Just return devicePath here.
+	glog.V(5).Infof("fc: AttachDisk volumeMode: %s", b.volumeMode)
```
single log, save a line

mtanino replied:
ack. done.

rootfs commented:
only reviewed FC part.

mtanino commented:
/retest

mtanino commented:
/retest

mtanino commented:
/retest

mtanino commented:
@rootfs 
I've addressed all your review comments. PTAL.
```
FC plugin updates  ... 151b3a7
- Remove unecessary comments
- Fix typos
- Return nil when ConstructBlockVolumeSpec failed to obtain global
  map path
```

rootfs commented:
thanks @mtanino FC part lgtm
/approve

mtanino commented:
/retest

mtanino commented:
/retest

mtanino commented:
@rootfs 
I've added one commit to address comments from @msau42.
If the FeatureGates is not enabled, function of block volume should be disabled. To do so, we need FeatureGates enablement check at both volumemanager and FC plugin.
PTAL.

```
FC plugin updates  …
- Add FeatureGates.Enabled checks to volumeMode and volumeDevices
```

Commit:
generated code for FC plugin change

mtanino commented:
@rootfs 
I've added two commits. PTAL on Monday.

```
FC plugin updates  … 31e483a
- Add FeatureGates.Enabled checks to volumeMode and volumeDevices

FC plugin update  … cf7b8ba
- Create spec from NewSpecFromPersistentVolume()
  Block Volume onlpy supports PV. We need to return
  spec with PV instead of volume
```

mtanino commented:
/retest

mtanino commented:
Removed API from this PR since it was merged.

mtanino commented:
/retest

rootfs commented:
@mtanino need a squash
@jsafrane any more comments?

mtanino commented:
@rootfs 

>> @mtanino need a squash
squash done for FC related commit. 
Volumemanager part will be removed once it lands. they need to pass unit tests for FC.

mtanino commented:
/retest

mtanino commented:
/sig storage

mtanino commented:
/retest

rootfs commented:
@mtanino cross build error
```
I1118 23:46:53.066] +++ [1118 23:45:41] windows/amd64: go build started
I1118 23:46:53.066] # k8s.io/kubernetes/cmd/kubelet/app
I1118 23:46:53.066] cmd/kubelet/app/server.go:417:49: too many arguments in call to cadvisor.New
I1118 23:46:53.066] 	have (string, uint, cadvisor.ImageFsInfoProvider, string, bool)
I1118 23:46:53.066] 	want (string, uint, cadvisor.ImageFsInfoProvider, string)
```

mtanino commented:
@rootfs 
pull-kubernetes-cross is hitting cross build failure. Not related to this PR.
See following.
`fix cadvisor.New signature for cross build #56003`

mtanino commented:
@rootfs @jsafrane @msau42 
All commits are squashed.

The volumemanager parts are synchronized between #51494 and this PR.
Once #51494 will get lgtm, we'll push both volumemanager fix and FC plugin fix via this PR.
It's fast path to merge both updates at the same time.
Is this ok?

I've added release note to this PR too. Also PR title and first comment of this PR need to be updated accordingly.

rootfs commented:
bazel build failed
```
W1008 15:21:57.229] 	pkg/kubelet/volumemanager/reconciler/reconciler.go: import of k8s.io/kubernetes/pkg/volume/util, which is not a direct dependency
W1008 15:21:57.314] ____Building complete.
```

mtanino commented:
/retest

e5e2fce was squashed already.
./hack/update-bazel.sh doesn't show any diff on my environment.
Not sure why this error reported today gain.

jsafrane commented:
/approve

rootfs commented:
/assign @random-liu

mtanino commented:
/retest

mtanino commented:
/retest

msau42 commented:
/kind feature
/priority important-soon

mtanino commented:
@rootfs @jsafrane @msau42 
Now I confirmed that all tests except cross-build are green. I'm going to remove volume manager core part from this PR.
```
pull-kubernetes-cross is hitting cross build failure. Not related to this PR.
See following.
fix cadvisor.New signature for cross build #56003
```

mtanino commented:
/retest

mtanino commented:
@rootfs @jsafrane @msau42 
I confirmed all tests exept kubernetes-cross were passed. I'm going to remove core volumemanager part from this PR. After this change, CI won't pass until #51494 gets land.
`Block volumes Support: CRI, volumemanager and operationexecutor changes #51494`

mtanino commented:
@rootfs @jsafrane 
Block Volume core PR is ready to merge. Please lgtm for this PR once it lands.

rootfs commented:
/lgtm

rootfs commented:
/retest

mtanino commented:
/retest

mtanino commented:
/test pull-kubernetes-verify
